h2o-core/src/main/java/water/Scope.java

```diff
@@ -20,56 +24,88 @@
  */
 
 public class Scope {
+  
+  private static final Logger log = LoggerFactory.getLogger(Scope.class);
+  
   // Thread-based Key lifetime tracking
-  static private final ThreadLocal<Scope> _scope = new ThreadLocal<Scope>() {
+  private static final ThreadLocal<Scope> _scope = new ThreadLocal<Scope>() {
     @Override protected Scope initialValue() { return new Scope(); }
   };
-  private final Stack<HashSet<Key>> _keys = new Stack<>();
+  private final Stack<Set<Key>> _keys = new Stack<>();
+  
+  private final Stack<Set<Key>> _protectedKeys = new Stack<>();
```

Comment:
@michalkurka you mentioned that you liked the concept but had questions regarding the implementation. Please share.